src/main/java/org/opensearch/security/configuration/SettingsPermissionValveImpl.java

```diff
@@ -0,0 +1,206 @@
+        }
+
+        // Validate transient settings
+        if (!validateSettingsMap(request.transientSettings(), allowedSettings)) {
+            auditLog.logMissingPrivileges(context.getAction(), request, context.getTask());
+            listener.onFailure(
+                new OpenSearchSecurityException("User not authorized to modify these cluster settings: " + request.transientSettings().keySet(), RestStatus.FORBIDDEN)
+            );
+            return false;
+        }
+
+        return true;
+    }
+
+    private boolean validateIndexSettings(
+        PrivilegesEvaluationContext context,
+        Settings requestSettings,
+        ActionRequest request,
+        ActionListener<?> listener
+    ) {
+        // Get allowed settings patterns from user's roles
+        Set<String> allowedSettings = getAllowedSettingsFromRoles(context);
+        // For backwards compatibility we will allow all settings if no allowed settings are defined
+        if (allowedSettings.isEmpty()) {
+            return true;
+        }
+        log.debug("Allowed settings: {} for user: {}", allowedSettings, context.getUser().getName());
+        if (!validateSettingsMap(requestSettings, allowedSettings)) {
+            auditLog.logMissingPrivileges(context.getAction(), request, context.getTask());
+            listener.onFailure(
+                new OpenSearchSecurityException("User not authorized to use these settings during index creation/update: " + requestSettings.keySet(), RestStatus.FORBIDDEN)
+            );
+            return false;
+        }
+
+        return true;
+    }
+
+    private Set<String> getAllowedSettingsFromRoles(PrivilegesEvaluationContext context) {
+        Set<String> allowedSettings = new HashSet<>();
+        SecurityDynamicConfiguration<RoleV7> roles = rolesConfiguration.get();
+
+        if (roles != null && roles.getCEntries() != null) {
+            for (String role : context.getUser().getRoles()) {
+                RoleV7 roleConfig = roles.getCEntries().get(role);
+                if (roleConfig != null) {
+                    // Get cluster-level settings permissions
+                    Set<String> clusterSettings = roleConfig.getAllowed_cluster_settings();
+                    if (clusterSettings != null) {
+                        allowedSettings.addAll(clusterSettings);
+                    }
+
+                    // Get index-level settings from index permissions
+                    List<RoleV7.Index> indexPermissions = roleConfig.getIndex_permissions();
+                    if (indexPermissions != null) {
+                        for (RoleV7.Index indexPermission : indexPermissions) {
+                            Set<String> indexSettings = indexPermission.getAllowed_settings();
+                            if (indexSettings != null) {
+                                allowedSettings.addAll(indexSettings);
+                            }
+                        }
+                    }
+                }
```

Comment:
I am not sure if I understand this correctly: We are merging here the settings from index permissions and cluster permissions. Don't we need to differentiate?